greedy_methods/kruskal.py

```diff
@@ -0,0 +1,120 @@
+"""
+https://en.wikipedia.org/wiki/Kruskal%27s_algorithm
+"""
+
+import doctest
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+
+@dataclass
+class Edge:
+    """
+    Represents an edge in a graph.
+
+    Attributes:
+        start (str): The starting vertex of the edge.
+        end (str): The ending vertex of the edge.
+        weight (int): The weight of the edge.
+    """
+
+    start: str
+    end: str
+    weight: int
+
+
+class DisjointSet:
+    """
+    Implements a disjoint-set (union-find) data structure.
+
+    This data structure is useful for tracking the partitioning of a set into disjoint,
+    non-overlapping subsets. It offers efficient operations for union and finding the
+    representative of a set.
+
+    Attributes:
+        parent (dict[str, str]): Maps each element to its parent element.
+        rank (dict[str, int]): Maps each element to its rank in the tree.
+
+    Methods:
+        find(item): Finds the representative of the set containing 'item'.
+        union(set1, set2): Merges the sets containing 'set1' and 'set2'.
+
+    >>> ds = DisjointSet(["A", "B", "C"])
+    >>> ds.find("A")
+    'A'
+    >>> ds.union("A", "B")
+    >>> ds.find("B")
+    'A'
+    >>> ds.union("B", "C")
+    >>> ds.find("C")
+    'A'
```

Comment:
Having all three answers be `'A'` does not raise confidence.

Let's use `disjoint_set` instead of `ds` as our variable name to make our code more self-documenting.

Please add tests where `disjoint_set = DisjointSet([])` and `disjoint_set = DisjointSet([1, 2, 3])`.

Let's add tests that fail like `>>> disjoint_set.find("D")`